src/intersection-observer-polyfill.js

```diff
@@ -0,0 +1,270 @@
+    // Check the element is an AMP element
+    dev().assert(element.isBuilt && element.isBuilt());
+    // Reset cached value for a new element
+
+    this.prevThresholdSlot_ = 0;
+    this.element_ = element;
+  }
+
+  // TODO: Support unobserve() function
+
+  /**
+   * Tick function that update the DOMRect of the root of observed element.
+   * Caller needs to make sure to pass in the correct container.
+   * Note: the opt_iframe param is the iframe position relative to the host doc,
+   * The iframe must be a non-scrollable iframe.
+   * @param {!./layout-rect.LayoutRectDef} viewport.
+   * @param {./layout-rect.LayoutRectDef=} opt_iframe.
+   */
+  tick(hostViewport, opt_iframe) {
+    if (opt_iframe) {
+      // If element inside an iframe. Adjust origin to the iframe.left/top.
+      hostViewport =
+          moveLayoutRect(hostViewport, -opt_iframe.left, -opt_iframe.top);
+      opt_iframe =
+          moveLayoutRect(opt_iframe, -opt_iframe.left, -opt_iframe.top);
+    }
+
+    // Normalize container LayoutRect to be relative to page
+    let elementRect;
+    let ownerRect = null;
+
+    // If calls with container, always assume getLayoutBox() return relative
+    // LayoutRect to container
+    elementRect = this.element_.getLayoutBox();
+    const owner = this.element_.getOwner();
+    ownerRect = owner && owner.getLayoutBox();
+
+    const changeEntry = this.getValidIntersectionChangeEntry_(
+        elementRect, ownerRect, hostViewport, opt_iframe);
+    if (changeEntry) {
+      this.callback_(changeEntry);
+    }
+  }
+
+  /**
+   * Return a change entry for that should be compatible with
+   * IntersectionObserverEntry if it's valid with current config.
+   * When the new intersection ratio doesn't cross one of a threshold value,
+   * the function will return null.
+   *
+   * Mutates passed in rootBounds to have x and y according to spec.
+   *
+   * @param {!./layout-rect.LayoutRectDef} element element's rect
+   * @param {?./layout-rect.LayoutRectDef} owner element's owner rect
+   * @param {!./layout-rect.LayoutRectDef} hostViewport hostViewport's rect
+   * @param {./layout-rect.LayoutRectDef=} opt_iframe. iframe container rect
+   * @return {?IntersectionObserverEntry} A valid change entry or null if ratio
+   * does not fill in a new threshold bucket.
+   * @private
+   */
+  getValidIntersectionChangeEntry_(element, owner, hostViewport, opt_iframe) {
+    // If opt_iframe is provided, all LayoutRect has position relative to
+    // the iframe.
+    // If opt_iframe is not provided, all LayoutRect has position relative to
+    // the host document.
+
+    // Building an IntersectionObserverEntry.
+    let intersectionRect = element;
+    const nonIntersectRect = layoutRectLtwh(0, 0, 0, 0);
+    // element intersects with its owner.
+    if (owner) {
+      intersectionRect = rectIntersection(owner, element) || nonIntersectRect;
+    }
+    // element intersects with container iframe.
+    if (opt_iframe) {
```

Comment:
Would it be the case: that the element has position absolute to the `opt_iframe` , thus outside `opt_iframe`, but intersects with the `hostViewport` ? I think calculating `rectIntersection` is still needed in this case. 
Relative test failure if I remove this check: https://github.com/ampproject/amphtml/pull/6100/files#diff-ff12640d23c90e5d1863b72bfc2e2a31R334